
syntax = "proto3";
package mimir;

message RegisterPartyReq { string party_id = 1; string addr = 2; bytes attestation_quote = 3; string attestation_kind = 4; }
message RegisterAck { bool ok = 1; string msg = 2; }

message CreateSessionReq { string session_id = 1; string model_name = 2; string mpc_ring = 3; }
message SessionAck { bool ok = 1; string msg = 2; }

message SubmitPromptReq { string session_id = 1; string owner_party = 2; repeated uint32 token_ids = 3; }
message SubmitPromptAck { bool ok = 1; string msg = 2; }

message RunInferenceReq { string session_id = 1; uint32 max_new_tokens = 2; uint32 topk = 3; }

message Token { uint32 token_id = 1; string text = 2; }
message StepTiming { uint32 step_index = 1; float ms = 2; }

message InferenceEvent {
  oneof ev {
    Token token = 1;
    StepTiming timing = 2;
  }
}

message Metrics { double tokens_per_second = 1; uint32 triple_pool_remaining = 2; }
message Empty {}

service Coordinator {
  rpc RegisterParty(RegisterPartyReq) returns (RegisterAck);
  rpc CreateSession(CreateSessionReq) returns (SessionAck);
  rpc SubmitPrompt(SubmitPromptReq) returns (SubmitPromptAck);
  rpc RunInference(RunInferenceReq) returns (stream InferenceEvent);
  rpc FetchMetrics(Empty) returns (Metrics);
}
